trieste/models/optimizer.py

```diff
@@ -0,0 +1,118 @@
+from typing import Any, Dict, Optional, Callable, Union, Iterable, Tuple
+from functools import singledispatch
+from dataclasses import dataclass, field
+import tensorflow as tf
+import gpflow
+
+from gpflow.models import ExternalDataTrainingLossMixin, InternalDataTrainingLossMixin
+from ..utils import jit
+from ..datasets import Dataset
+
+
+DatasetBuilder = Callable[[Dataset, Optional[int]], Union[Iterable, Tuple]]
+"""
+Type alias for a function that creates minibatches from a :class:`~optiflow.datasets.Dataset`.
+"""
+
+
+@dataclass
+class Optimizer:
+    optimizer: Union[gpflow.optimizers.Scipy, tf.optimizers.Optimizer]
+    minimize_args: Dict[str, Any] = field(default_factory=lambda: {})
+    compile: bool = False
+
+    def create_loss(self, model: tf.Module, dataset: Dataset) -> Callable:
+        x = tf.convert_to_tensor(dataset.query_points)
+        y = tf.convert_to_tensor(dataset.observations)
+        data = (x, y)
+        return create_loss_function(model, data, self.compile)
+
+    def optimize(self, model: tf.Module, dataset: Dataset):
+        loss_fn = self.create_loss(model, dataset)
+        variables = model.trainable_variables
+        return self.optimizer.minimize(loss_fn, variables, **self.minimize_args)
+
+
+@dataclass
+class TFOptimizer(Optimizer):
+    max_iter: int = 100
+    batch_size: Optional[int] = None
+    dataset_builder: Optional[DatasetBuilder] = None
+
+    def __post_init__(self):
+        if self.max_iter is None:
+            raise ValueError("Max iterations expected to be set")
+
+    def create_loss(self, model: tf.Module, dataset: Dataset) -> Callable:
+        def creator_fn(data: Union[Tuple, Iterable]):
+            return create_loss_function(model, data, self.compile)
+
+        if self.dataset_builder is None and self.batch_size is None:
+            x = tf.convert_to_tensor(dataset.query_points)
+            y = tf.convert_to_tensor(dataset.observations)
+            return creator_fn((x, y))
+        elif self.dataset_builder is None:
+            d = tf.data.Dataset.from_tensor_slices((dataset.query_points, dataset.observations))
+            size = len(dataset)
+            data: Iterable = iter(
+                d.shuffle(size)
+                .batch(self.batch_size)
+                .prefetch(tf.data.experimental.AUTOTUNE)
+                .repeat()
+            )
+            return creator_fn(data)
```

Comment:
this looks like it can be written as a default `dataset_builder`, in which case the `batch_size` can be baked into the `dataset_builder`, and isn't needed on this class, which feels intuitive to me: a dataset builder builds minibatches from a Dataset. This process can be defined without the notion of a batch size

Reply:
The `batch_size` is the only variable that changes here, the user is not obliged to write his own builder all the time. It is enough to set `batch_size` to an appropriate value.